Open Judge System/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs

```diff
@@ -0,0 +1,134 @@
+    using OJS.Services.Data.Participants;
+
+    public class ParticipantScoresDataService : IParticipantScoresDataService
+    {
+        private readonly IEfGenericRepository<ParticipantScore> participantScores;
+        private readonly IParticipantsDataService participantsData;
+
+        public ParticipantScoresDataService(
+            IEfGenericRepository<ParticipantScore> participantScores,
+            IParticipantsDataService participantsData)
+        {
+            this.participantScores = participantScores;
+            this.participantsData = participantsData;
+        }
+
+        public ParticipantScore GetByParticipantIdAndProblemId(int participantId, int problemId) =>
+            this.participantScores
+                .All()
+                .FirstOrDefault(ps => ps.ParticipantId == participantId &&
+                    ps.ProblemId == problemId);
+
+        public ParticipantScore GetByParticipantIdProblemIdAndIsOfficial(int participantId, int problemId, bool isOfficial) =>
+            this.participantScores
+                .All()
+                .FirstOrDefault(ps => ps.ParticipantId == participantId &&
+                    ps.ProblemId == problemId &&
+                    ps.IsOfficial == isOfficial);
+
+        public IQueryable<ParticipantScore> GetAllQuery() =>
+            this.participantScores.All();
+
+        public void SaveBySubmission(Submission submission, bool resetScore = false)
+        {
+            if (submission.ParticipantId == null || submission.ProblemId == null)
+            {
+                return;
+            }
+
+            var participant = this.participantsData
+                .GetByIdQuery(submission.ParticipantId.Value)
+                .Select(p => new
+                {
+                    p.IsOfficial,
+                    p.User.UserName
+                })
+                .FirstOrDefault();
+
+            if (participant != null)
+            {
+                using (var transaction = this.participantScores.BeginTransaction())
+                {
+                    var existingScore = this.GetByParticipantIdProblemIdAndIsOfficial(
+                        submission.ParticipantId.Value,
+                        submission.ProblemId.Value,
+                        participant.IsOfficial);
+
+                    if (existingScore == null)
+                    {
+                        this.AddNew(submission, participant.UserName, participant.IsOfficial);
+                    }
+                    else if (resetScore ||
+                        submission.Points >= existingScore.Points ||
+                        submission.Id == existingScore.SubmissionId)
+                    {
+                        this.Update(existingScore, submission.Id, submission.Points);
+                    }
+
+                    transaction.Commit();
+                }
+            }
+        }
+
+        public void DeleteAllByProblem(int problemId) =>
+            this.participantScores.All()
```

Comment:
`.All()` on new line or place the other expressions on the same line if < 120 characters